spi/core-spi/src/main/java/org/eclipse/dataspaceconnector/spi/result/AbstractResult.java

```diff
@@ -68,4 +70,56 @@ public boolean failed() {
     public String getFailureDetail() {
         return failure == null ? null : String.join(", ", getFailureMessages());
     }
+
+    /**
+     * Executes a {@link Consumer} if this {@link Result} is successful
+     */
+    public void ifSuccess(Consumer<T> successAction) {
```

Comment:
how consolidating `ifSuccess` and `whenSuccessful` by adding an overload for `whenSuccessful`:
```java
public AbstractResult<T, F> whenSuccessful(Consumer<AbstractResult<T, F>> content) {...}
public AbstractResult<T, F> whenSuccessful(Consumer<T> successAction) {...}
```
Conversely, we can indeed return `this` from the `ifFailure` method, as it is semantically different from `orElse`. Those two would be very similar, except for the method name.

Reply:
to be honest I have some opinions about the `Optional` interface :)
the `whenSuccessful` is different since the action get the whole result (that's probably not needed).
I would keep only one method that will execute an action on success and one that will do the same on failure.

Why do not call them `onSuccess` and `onFailure`?

Reply:
maybe we can even simplify it a bit:
- have one method each for success and failure (FWIW lets call them `onSuccess` and `onFailure`, as you suggested), receiving either the `content` or the `Failure`, returning `this`
- have one method called `flatMap` that accepts a mapping function and converts one Result into another. This would get rid of the `whenSuccess(ful)` stuff for now

that way, we can see if that fits the bill and extend later.